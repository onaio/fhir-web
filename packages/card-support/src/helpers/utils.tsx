/** Function to download data to a file
 *
 * @param {string} data - data to be written to file
 * @param {string} filename - name of the file to be saved
 * @param {string} type - MIME type for the file
 */
export const downloadFile = (data: string, filename: string, type: string) => {
  const file = new Blob([data], { type });
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(file, filename);
  } else {
    // Others
    const docElement = document.createElement('a');
    const url = URL.createObjectURL(file);
    docElement.href = url;
    docElement.download = filename;
    document.body.appendChild(docElement);
    docElement.click();
    setTimeout(() => {
      document.body.removeChild(docElement);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
};
