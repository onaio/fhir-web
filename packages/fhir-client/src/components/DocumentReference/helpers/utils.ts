import { get } from 'lodash';

/**
 * get object from an array, the check parameters are passed in as an obj, keys in keyValue can represent a path
 *
 * @param objArray - array obj with records of objs
 * @param keyValue - obj defining what key and values to use when matching
 * @param all - return first matched or all
 */
export function findObj<T extends object>(
  objArray?: T[],
  keyValue?: Record<keyof T | string, unknown>,
  all = false
) {
  if (objArray && keyValue) {
    const checkParameters = Object.entries(keyValue);
    const matched = objArray.filter((obj) => {
      let matches = true;
      checkParameters.forEach(([key, value]) => {
        if (get(obj, key) !== value) {
          matches = false;
        }
      });
      return matches;
    });
    if (all) {
      return matched;
    } else {
      return matched[0];
    }
  }
  if (!objArray || !keyValue) return;
}
