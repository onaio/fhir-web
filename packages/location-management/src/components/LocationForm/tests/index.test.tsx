import { store } from '@opensrp/store';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { RouteComponentProps, Router } from 'react-router';
import { LocationForm } from '..';
import { defaultFormField, FormInstances, getLocationFormFields } from '../utils';
import { createBrowserHistory } from 'history';
import { authenticateUser } from '@onaio/session-reducer';
import { Form } from 'antd';
import {
  createdLocation1,
  duplicateLocationTags,
  fetchCalls1,
  generatedLocation2,
  generatedLocation4,
  generatedLocation4Dot1,
  location2,
  location4,
  locationSettings,
  locationTags,
  rawOpenSRPHierarchy1,
  serviceTypeSettings,
} from './fixtures';

const history = createBrowserHistory();

jest.mock('uuid', () => {
  const actual = jest.requireActual('uuid');
  return {
    ...actual,
    v4: () => '9b782015-8392-4847-b48c-50c11638656b',
  };
});

// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('jest-fetch-mock');

describe('LocationForm', () => {
  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        // eslint-disable-next-line @typescript-eslint/camelcase
        { api_token: 'hunter2', oAuth2Data: { access_token: 'sometoken', state: 'abcde' } }
      )
    );
  });

  afterEach(() => {
    fetch.resetMocks();
  });

  it('renders correctly', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    fetch.mockResponse(JSON.stringify([]));

    const wrapper = mount(
      <Router history={history}>
        <LocationForm />
      </Router>,
      { attachTo: div }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // getter fetch calls made
    expect(fetch.mock.calls).toEqual(fetchCalls1);

    expect(toJson(wrapper.find('#instance label'))).toMatchSnapshot('instance label');
    expect(toJson(wrapper.find('#instance input'))).toMatchSnapshot('instance field');

    expect(toJson(wrapper.find('#parentId label'))).toMatchSnapshot('parentId label');
    expect(toJson(wrapper.find('#parentId select'))).toMatchSnapshot('parentId field');

    expect(toJson(wrapper.find('#name label'))).toMatchSnapshot('name label');
    expect(toJson(wrapper.find('#name input'))).toMatchSnapshot('name field');

    expect(toJson(wrapper.find('#status label').first())).toMatchSnapshot('status label');
    expect(toJson(wrapper.find('#status input'))).toMatchSnapshot('status field');

    expect(toJson(wrapper.find('#isJurisdiction label').first())).toMatchSnapshot(
      'isJurisdiction label'
    );
    expect(toJson(wrapper.find('#isJurisdiction input'))).toMatchSnapshot('isJurisdiction field');
    expect(toJson(wrapper.find('#type label'))).toMatchSnapshot('type label');
    expect(toJson(wrapper.find('#type input'))).toMatchSnapshot('type field');
    expect(toJson(wrapper.find('#serviceType label'))).toMatchSnapshot('serviceType label');
    expect(toJson(wrapper.find('#serviceType input'))).toMatchSnapshot('serviceType field');
    expect(toJson(wrapper.find('#externalId label'))).toMatchSnapshot('externalId label');
    expect(toJson(wrapper.find('#externalId input'))).toMatchSnapshot('externalId field');
    expect(toJson(wrapper.find('#geometry label'))).toMatchSnapshot('geometry label');
    expect(toJson(wrapper.find('#geometry textarea'))).toMatchSnapshot('geometry field');
    expect(toJson(wrapper.find('#longitude label'))).toMatchSnapshot('longitude label');
    expect(toJson(wrapper.find('#longitude input'))).toMatchSnapshot('longitude field');
    expect(toJson(wrapper.find('#latitude label'))).toMatchSnapshot('latitude label');
    expect(toJson(wrapper.find('#latitude input'))).toMatchSnapshot('latitude field');
    expect(toJson(wrapper.find('#locationTags label'))).toMatchSnapshot('locationTags label');
    expect(toJson(wrapper.find('#locationTags input'))).toMatchSnapshot('locationTags field');

    wrapper.find('.extraFields').forEach((field) => {
      expect(toJson(field.find('label'))).toMatchSnapshot('field label');
      expect(toJson(field.find('input'))).toMatchSnapshot('field input');
    });

    expect(toJson(wrapper.find('#location-form-submit-button button'))).toMatchSnapshot(
      'submit button'
    );
    expect(toJson(wrapper.find('#location-form-cancel-button button'))).toMatchSnapshot(
      'cancel button'
    );
    wrapper.unmount();
  });

  it('form validation works for core instance', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    fetch.mockResponse(JSON.stringify([]));

    // when instance is set to core by default, types is required, serviceType is not required
    const wrapper = mount(
      <Router history={history}>
        <LocationForm />
      </Router>,
      { attachTo: div }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('FormItem#instance').text()).toMatchInlineSnapshot(`"Instance"`);

    expect(wrapper.find('FormItem#parentId').text()).toMatchInlineSnapshot(
      `"ParentSelect the parent location"`
    );

    expect(wrapper.find('FormItem#status').text()).toMatchInlineSnapshot(`"StatusActiveInactive"`);

    expect(wrapper.find('FormItem#isJurisdiction').text()).toMatchInlineSnapshot(
      `"Location categoryService pointJurisdiction"`
    );

    // type is required for core
    expect(wrapper.find('FormItem#type').text()).toMatchInlineSnapshot(
      `"TypeType can only contain letters, numbers and spaces"`
    );

    // name is required for core
    expect(wrapper.find('FormItem#name').text()).toMatchInlineSnapshot(`"NameName is required"`);

    // not required for core
    expect(wrapper.find('FormItem#serviceType').text()).toMatchInlineSnapshot(`"Type"`);

    expect(wrapper.find('FormItem#externalId').text()).toMatchInlineSnapshot(`"External ID"`);

    expect(wrapper.find('FormItem#geometry').text()).toMatchInlineSnapshot(`"Geometry"`);

    expect(wrapper.find('FormItem#locationTags').text()).toMatchSnapshot(
      'location Tags does not have error message'
    );

    wrapper.unmount();
  });

  it('form validation works for wrong data types', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    fetch.mockResponse(JSON.stringify([]));

    const wrapper = mount(
      <Router history={history}>
        <LocationForm />
      </Router>,
      { attachTo: div }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // set longitude, and latitude to invalid values
    wrapper
      .find('FormItem#longitude input')
      .simulate('change', { target: { value: '432dsff', name: 'longitude' } });

    wrapper
      .find('FormItem#latitude input')
      .simulate('change', { target: { value: '43f', name: 'latitude' } });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('FormItem#longitude').text()).toMatchInlineSnapshot(
      `"LongitudeOnly decimal values allowed"`
    );

    expect(wrapper.find('FormItem#latitude').text()).toMatchInlineSnapshot(
      `"LatitudeOnly decimal values allowed"`
    );

    wrapper.unmount();
  });

  it('form validation works for eusm instance', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    fetch.mockResponse(JSON.stringify([]));

    // when instance is set to core by default, types is required, serviceType is not required
    const initialValues = { ...defaultFormField, instance: FormInstances.EUSM };
    const wrapper = mount(
      <Router history={history}>
        <LocationForm initialValues={initialValues} />
      </Router>,
      {
        attachTo: div,
      }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('FormItem#instance').text()).toMatchInlineSnapshot(`"Instance"`);

    expect(wrapper.find('FormItem#parentId').text()).toMatchInlineSnapshot(
      `"ParentSelect the parent location'parentId' is required"`
    );

    expect(wrapper.find('FormItem#status').text()).toMatchInlineSnapshot(`"StatusActiveInactive"`);

    expect(wrapper.find('FormItem#isJurisdiction').text()).toMatchInlineSnapshot(
      `"Location categoryService pointJurisdiction"`
    );

    // type is required for core, but not required for eusm instance
    expect(wrapper.find('FormItem#type').text()).toMatchInlineSnapshot(`"Type"`);

    // name is required for core
    expect(wrapper.find('FormItem#name').text()).toMatchInlineSnapshot(`"NameName is required"`);

    // service types is required for eusm
    expect(wrapper.find('FormItem#serviceType').text()).toMatchInlineSnapshot(
      `"TypeService types is required"`
    );

    expect(wrapper.find('FormItem#externalId').text()).toMatchInlineSnapshot(`"External ID"`);

    expect(wrapper.find('FormItem#geometry').text()).toMatchInlineSnapshot(`"Geometry"`);

    expect(wrapper.find('FormItem#locationTags').text()).toMatchSnapshot(
      'location Tags does not have error message'
    );
    wrapper.unmount();
  });

  it('submits new location unit', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch.mockResponse(JSON.stringify([]));

    const wrapper = mount(
      <Router history={history}>
        <LocationForm />
      </Router>,
      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formInstance = (wrapper.find(Form).props() as any).form;

    // change parent Id
    formInstance.setFieldsValue({
      parentId: '51',
    });

    // simulate active check to be inactive
    wrapper
      .find('FormItem#status input')
      .last()
      .simulate('change', {
        target: { checked: true },
      });

    // set isJurisdiction to false
    wrapper
      .find('FormItem#isJurisdiction input')
      .first()
      .simulate('change', {
        target: { checked: true },
      });

    // simulate type field change
    wrapper
      .find('FormItem#type input')
      .simulate('change', { target: { name: 'type', value: 'Feature' } });

    // simulate name change
    wrapper
      .find('FormItem#name input')
      .simulate('change', { target: { name: 'name', value: 'area51' } });

    // simulate service type change
    // change service types
    formInstance.setFieldsValue({
      serviceType: 'School',
    });

    wrapper
      .find('FormItem#externalId input')
      .simulate('change', { target: { name: 'externalId', value: 'secret' } });

    wrapper.find('FormItem#geometry textarea').simulate('change', {
      target: { value: JSON.stringify([19.92919921875, 30.135626231134587]) },
    });

    fetch.resetMocks();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(fetch.mock.calls).toEqual([
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=false',
        {
          'Cache-Control': 'no-cache',
          Pragma: 'no-cache',
          body: JSON.stringify(createdLocation1),
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'POST',
        },
      ],
    ]);

    wrapper.unmount();
  });

  it('correctly redirects on submit', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch.mockResponse(JSON.stringify([]));

    const someMockURL = '/someURL';
    const successURLGeneratorMock = jest.fn(() => someMockURL);

    const wrapper = mount(
      <Router history={history}>
        <LocationForm successURLGenerator={successURLGeneratorMock} />
      </Router>,
      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formInstance = (wrapper.find(Form).props() as any).form;

    // change parent Id
    formInstance.setFieldsValue({
      parentId: '51',
    });

    // simulate active check to be inactive
    wrapper
      .find('FormItem#status input')
      .last()
      .simulate('change', {
        target: { checked: true },
      });

    // set isJurisdiction to false
    wrapper
      .find('FormItem#isJurisdiction input')
      .first()
      .simulate('change', {
        target: { checked: true },
      });

    // simulate type field change
    wrapper
      .find('FormItem#type input')
      .simulate('change', { target: { name: 'type', value: 'Feature' } });

    // simulate name change
    wrapper
      .find('FormItem#name input')
      .simulate('change', { target: { name: 'name', value: 'area51' } });

    // simulate service type change
    // change service types
    formInstance.setFieldsValue({
      serviceType: 'School',
    });

    wrapper
      .find('FormItem#externalId input')
      .simulate('change', { target: { name: 'externalId', value: 'secret' } });

    wrapper.find('FormItem#geometry textarea').simulate('change', {
      target: { value: JSON.stringify([19.92919921875, 30.135626231134587]) },
    });

    fetch.resetMocks();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(successURLGeneratorMock).toHaveBeenCalledWith(createdLocation1);
    expect(
      (wrapper.find('Router').props() as RouteComponentProps).history.location.pathname
    ).toEqual(someMockURL);
    wrapper.unmount();
  });

  it('cancel handler is called on cancel', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch.mockResponse(JSON.stringify([]));
    const cancelMock = jest.fn();

    const wrapper = mount(
      <Router history={history}>
        <LocationForm onCancel={cancelMock} />
      </Router>,
      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('button#location-form-cancel-button').simulate('click');
    wrapper.update();

    expect(cancelMock).toHaveBeenCalled();
    wrapper.unmount();
  });

  it('is able to edit a location unit', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch
      .once(JSON.stringify([location2]))
      .once(JSON.stringify(serviceTypeSettings))
      .once(JSON.stringify(locationTags))
      .once(JSON.stringify(locationSettings))
      .once(JSON.stringify(rawOpenSRPHierarchy1));

    const initialValues = getLocationFormFields(location2);

    const locationFormProps = {
      initialValues,
    };

    const wrapper = mount(
      <Router history={history}>
        <LocationForm {...locationFormProps} />
      </Router>,
      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // put this here to help guide to know how to mock the fetch calls
    expect(fetch.mock.calls.map((call) => call[0])).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location/findByProperties?is_jurisdiction=true&return_geometry=false&properties_filter=status:Active,geographicLevel:0',
      'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings/?serverVersion=0&identifier=service_point_types',
      'https://opensrp-stage.smartregister.org/opensrp/rest/location-tag',
      'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings/?serverVersion=0&identifier=location_settings',
      'https://opensrp-stage.smartregister.org/opensrp/rest/location/hierarchy/95310ca2-02df-47ba-80fc-bf31bfaa88d7?return_structure_count=false',
    ]);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formInstance = (wrapper.find(Form).props() as any).form;

    // simulate active check to be inactive
    wrapper
      .find('FormItem#status input')
      .last()
      .simulate('change', {
        target: { checked: true },
      });

    // simulate name change
    wrapper
      .find('FormItem#name input')
      .simulate('change', { target: { name: 'name', value: 'Mars' } });

    // simulate service type change
    // change service types
    formInstance.setFieldsValue({
      serviceType: 'School',
    });

    wrapper
      .find('FormItem#externalId input')
      .simulate('change', { target: { name: 'externalId', value: 'alien' } });

    const geometry = {
      type: 'Point',
      coordinates: [19.92919921875, 30.135626231134587],
    };

    wrapper.find('FormItem#geometry textarea').simulate('change', {
      target: { value: JSON.stringify(geometry) },
    });

    // extra fields
    expect(wrapper.find('FormItem.extra-fields')).toHaveLength(4);
    wrapper.find('FormItem.extra-fields').forEach((field) => {
      // snapshot of label
      expect(toJson(field.find('label'))).toMatchSnapshot('field label');
      expect(toJson(field.find('input'))).toMatchSnapshot('field input');

      // fill the fields
      field
        .find('input')
        .simulate('change', { target: { value: `extraFields - ${field.find('label').text()}` } });
    });

    fetch.resetMocks();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(fetch.mock.calls).toEqual([
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
        {
          'Cache-Control': 'no-cache',
          Pragma: 'no-cache',
          body: JSON.stringify(generatedLocation2),
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'PUT',
        },
      ],
    ]);

    wrapper.unmount();
  });

  it('Editing latitude and longitudes works fine', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch
      .once(JSON.stringify([location2]))
      .once(JSON.stringify(serviceTypeSettings))
      .once(JSON.stringify(locationTags))
      .once(JSON.stringify(locationSettings))
      .once(JSON.stringify(rawOpenSRPHierarchy1));

    const initialValues = getLocationFormFields(location4);

    const locationFormProps = {
      initialValues,
    };

    const wrapper = mount(
      <Router history={history}>
        <LocationForm {...locationFormProps} />
      </Router>,

      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formInstance = (wrapper.find(Form).props() as any).form;

    // simulate service type change
    // change service types
    formInstance.setFieldsValue({
      serviceType: 'School',
    });

    wrapper.find('FormItem#latitude input').simulate('change', {
      target: { value: '34.56' },
    });

    wrapper.find('FormItem#longitude input').simulate('change', {
      target: { value: '19.56' },
    });

    fetch.resetMocks();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(fetch.mock.calls).toEqual([
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
        {
          'Cache-Control': 'no-cache',
          Pragma: 'no-cache',
          body: JSON.stringify(generatedLocation4),
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'PUT',
        },
      ],
    ]);
    wrapper.unmount();
  });

  it('#595 Duplicate location Tags failing upload', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    fetch
      .once(JSON.stringify([location2]))
      .once(JSON.stringify(serviceTypeSettings))
      .once(JSON.stringify(duplicateLocationTags))
      .once(JSON.stringify(locationSettings))
      .once(JSON.stringify(rawOpenSRPHierarchy1));

    const initialValues = getLocationFormFields(location4);

    const locationFormProps = {
      initialValues,
    };

    const wrapper = mount(
      <Router history={history}>
        <LocationForm {...locationFormProps} />
      </Router>,

      { attachTo: container }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    fetch.mockReset();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    /** payload does not contain duplicate entries in locationTags field*/
    expect(generatedLocation4Dot1.locationTags).toHaveLength(1);
    expect(fetch.mock.calls).toEqual([
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
        {
          'Cache-Control': 'no-cache',
          Pragma: 'no-cache',
          body: JSON.stringify(generatedLocation4Dot1),
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'PUT',
        },
      ],
    ]);
    wrapper.unmount();
  });
});
