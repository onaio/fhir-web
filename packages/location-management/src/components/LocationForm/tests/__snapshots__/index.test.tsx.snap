// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationForm form validation works for core instance: location Tags does not have error message 1`] = `"Unit group Enter a location group name"`;

exports[`LocationForm form validation works for eusm instance: location Tags does not have error message 1`] = `"Unit group Enter a location group name"`;

exports[`LocationForm is able to edit a location unit: field input 1`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_extraFields_0_Sample Key"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value=""
/>
`;

exports[`LocationForm is able to edit a location unit: field input 2`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_extraFields_1_Area Nick name"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value=""
/>
`;

exports[`LocationForm is able to edit a location unit: field input 3`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_extraFields_2_Sample Key Two"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value=""
/>
`;

exports[`LocationForm is able to edit a location unit: field input 4`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_extraFields_3_Area code"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value=""
/>
`;

exports[`LocationForm is able to edit a location unit: field label 1`] = `
<label
  className=""
  htmlFor="location-form_extraFields_0_Sample Key"
  title="Sample Key"
>
  Sample Key
</label>
`;

exports[`LocationForm is able to edit a location unit: field label 2`] = `
<label
  className=""
  htmlFor="location-form_extraFields_1_Area Nick name"
  title="Area Nick name"
>
  Area Nick name
</label>
`;

exports[`LocationForm is able to edit a location unit: field label 3`] = `
<label
  className=""
  htmlFor="location-form_extraFields_2_Sample Key Two"
  title="Sample Key Two"
>
  Sample Key Two
</label>
`;

exports[`LocationForm is able to edit a location unit: field label 4`] = `
<label
  className=""
  htmlFor="location-form_extraFields_3_Area code"
  title="Area code"
>
  Area code
</label>
`;

exports[`LocationForm renders correctly: cancel button 1`] = `
<button
  className="ant-btn"
  id="location-form-cancel-button"
  onClick={[Function]}
  type="button"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Cancel
  </span>
</button>
`;

exports[`LocationForm renders correctly: externalId field 1`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_externalId"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Select status"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: externalId label 1`] = `
<label
  className=""
  htmlFor="location-form_externalId"
  title="External ID"
>
  External ID
</label>
`;

exports[`LocationForm renders correctly: geometry field 1`] = `
<textarea
  className="ant-input"
  disabled={false}
  id="location-form_geometry"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="</> JSON"
  rows={4}
  style={Object {}}
  value=""
/>
`;

exports[`LocationForm renders correctly: geometry label 1`] = `
<label
  className=""
  htmlFor="location-form_geometry"
  title="Geometry"
>
  Geometry
</label>
`;

exports[`LocationForm renders correctly: instance field 1`] = `
<input
  className="ant-input ant-input-disabled"
  disabled={true}
  id="location-form_instance"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value="core"
/>
`;

exports[`LocationForm renders correctly: instance label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_instance"
  title="Instance"
>
  Instance
</label>
`;

exports[`LocationForm renders correctly: isJurisdiction field 1`] = `
Array [
  <input
    checked={false}
    className="ant-radio-input"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value={false}
  />,
  <input
    checked={true}
    className="ant-radio-input"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value={true}
  />,
]
`;

exports[`LocationForm renders correctly: isJurisdiction label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_isJurisdiction"
  title="Location category"
>
  Location category
</label>
`;

exports[`LocationForm renders correctly: locationTags field 1`] = `
<input
  aria-activedescendant="location-form_locationTags_list_0"
  aria-autocomplete="list"
  aria-controls="location-form_locationTags_list"
  aria-haspopup="listbox"
  aria-owns="location-form_locationTags_list"
  autoComplete="off"
  className="ant-select-selection-search-input"
  disabled={false}
  id="location-form_locationTags"
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  readOnly={true}
  role="combobox"
  style={
    Object {
      "opacity": 0,
    }
  }
  type="search"
  unselectable="on"
  value=""
/>
`;

exports[`LocationForm renders correctly: locationTags label 1`] = `
<label
  className=""
  htmlFor="location-form_locationTags"
  title="Unit group"
>
  Unit group
</label>
`;

exports[`LocationForm renders correctly: name field 1`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_name"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Enter a location name"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: name label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_name"
  title="Name"
>
  Name
</label>
`;

exports[`LocationForm renders correctly: parentId field 1`] = `null`;

exports[`LocationForm renders correctly: parentId label 1`] = `
<label
  className=""
  htmlFor="location-form_parentId"
  title="Parent"
>
  Parent
</label>
`;

exports[`LocationForm renders correctly: serviceType field 1`] = `
<input
  aria-activedescendant="location-form_serviceType_list_0"
  aria-autocomplete="list"
  aria-controls="location-form_serviceType_list"
  aria-haspopup="listbox"
  aria-owns="location-form_serviceType_list"
  autoComplete="off"
  className="ant-select-selection-search-input"
  disabled={false}
  id="location-form_serviceType"
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  readOnly={true}
  role="combobox"
  style={
    Object {
      "opacity": 0,
    }
  }
  type="search"
  unselectable="on"
  value=""
/>
`;

exports[`LocationForm renders correctly: serviceType label 1`] = `
<label
  className=""
  htmlFor="location-form_serviceType"
  title="Type"
>
  Type
</label>
`;

exports[`LocationForm renders correctly: status field 1`] = `
Array [
  <input
    checked={true}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="Active"
  />,
  <input
    checked={false}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="InActive"
  />,
]
`;

exports[`LocationForm renders correctly: status label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_status"
  title="Status"
>
  Status
</label>
`;

exports[`LocationForm renders correctly: submit button 1`] = `
<button
  className="ant-btn ant-btn-primary"
  disabled={false}
  id="location-form-submit-button"
  onClick={[Function]}
  type="submit"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Save
  </span>
</button>
`;

exports[`LocationForm renders correctly: type field 1`] = `
<input
  className="ant-input"
  disabled={false}
  id="location-form_type"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Select type"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: type label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_type"
  title="Type"
>
  Type
</label>
`;
